be/src/util/cpu_usage_info.cpp

```diff
@@ -0,0 +1,61 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021 StarRocks Limited.
+
+#include "util/cpu_usage_info.h"
+
+#include <glog/logging.h>
+
+#include <cstdio>
+#include <thread>
+
+#include "util/defer_op.h"
+#include "util/time.h"
+
+namespace starrocks {
+
+const int CpuUsageRecorder::NUM_HARDWARE_CORES = std::thread::hardware_concurrency();
+const int CpuUsageRecorder::SECOND_CLOCK_TICK = sysconf(_SC_CLK_TCK);
+
+CpuUsageRecorder::CpuUsageRecorder() : _timestamp{0, MonotonicNanos()}, _proc_time{0, _get_proc_time()} {}
+
+CpuUsageRecorder::~CpuUsageRecorder() {
+    if (_line_ptr != nullptr) {
+        free(_line_ptr);
+    }
+}
+
+void CpuUsageRecorder::update_interval() {
+    _curr_idx = (_curr_idx + 1) % 2;
+    _proc_time[_curr_idx] = _get_proc_time();
+    _timestamp[_curr_idx] = MonotonicNanos();
+}
+
+int CpuUsageRecorder::cpu_used_permille() const {
+    if (_curr_idx == ABSENT_INDEX) {
+        // Haven't called update_interval.
+        return 0;
+    }
+    int prev_idx = (_curr_idx + 1) % 2;
+    return (_proc_time[_curr_idx] - _proc_time[prev_idx]) * 1000'000'000L / SECOND_CLOCK_TICK * 1000 /
+           NUM_HARDWARE_CORES / (_timestamp[_curr_idx] - _timestamp[prev_idx]);
+}
+
+uint64_t CpuUsageRecorder::_get_proc_time() {
+    FILE* fp = fopen("/proc/self/stat", "r");
```

Comment:
pls consider different OS

Reply:
For now, we only support Linux OS, and `/proc/self/stat` is supported by all the Linux version (see [main-proc](https://man7.org/linux/man-pages/man5/proc.5.html)).
BTW, we use lots of `/proc/xxx` in codes, including disk metric, cpu metric, and other system metrics.

Reply:
What if multi be processes are in the same machine? or the be process is in the Docker(eg. compute/storage disaggregaed)?

Reply:
1. Multiple BE processes cannot run in the same machine. Otherwise, how can we isolate the resource usage between them?
2. `/proc/self/stat` is also effective for processes in Docker, but `std::thread::hardware_concurrency` returns the number of the total CPUs in the host machine not `cpuset.cpus`. However, this is an existed problem, since we use `std::thread::hardware_concurrency` in lots of codes. I would fix all the usages of `hardware_concurrency` by considering `cpuset.cpus` in the next PR.